### PR TITLE
ad controller: optimize populateActualStateOfWorld

### PR DESCRIPTION
**Pull request by huww98**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

ad controller: optimize populateActualStateOfWorld

Added benchmark for this optimization. Fixed the broken unit test. And increased the unit test coverage of relevant code.

Reduce overall complexity of `populateActualStateOfWorld` from O(n^2) to O(n)
Run time of the new benchmark is reduced from hours to 13s (among them, 8s is in logging)

The original version is so slow, it takes 59s to process 500 nodes. It should take about 400 minutes for 10000 nodes from calculation. I cancel the benchmark after 10 minutes.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by huww98** [comment]
Failing test possibly the same as #121233
```
=== RUN   TestWebSocketClient_TextMessageTypeError
E1028 03:35:05.544138   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:35000->127.0.0.1:36191: use of closed network connection
E1028 03:35:06.524954   65205 conn.go:347] Unable to write frame to 0: io: read/write on closed pipe
◈�Z'7��y�}�ӌz��/��_�S���k�@vв���	g��.x'��q��y�'�)��#+H����`�����4YO��+(z��|��Ձ�^��������Y؆ZWu�ژ2��u��=�mu��@I���z�ެ�c�
E1028 03:35:06.525103   65205 conn.go:315] Error on socket receive: read tcp 127.0.0.1:36191->127.0.0.1:35000: use of closed network connection
E1028 03:35:10.518668   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:47448->127.0.0.1:35513: use of closed network connection
E1028 03:35:10.519776   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:47456->127.0.0.1:35513: use of closed network connection
E1028 03:35:10.520883   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:47458->127.0.0.1:35513: use of closed network connection
E1028 03:35:15.532625   65205 websocket.go:486] Websocket Ping failed: write tcp 127.0.0.1:52250->127.0.0.1:33223: write: broken pipe
    websocket_test.go:931: expect stream to be closed after connection is closed.
--- FAIL: TestWebSocketClient_TextMessageTypeError (30.00s)
```
/test pull-kubernetes-unit

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 1]
```diff
@@ -384,6 +384,11 @@ func (adc *attachDetachController) populateActualStateOfWorld(logger klog.Logger
 
 	for _, node := range nodes {
 		nodeName := types.NodeName(node.Name)
+		volumesInUse := make(map[v1.UniqueVolumeName]struct{}, len(node.Status.VolumesInUse))
+		for _, volumeName := range node.Status.VolumesInUse {
+			volumesInUse[volumeName] = struct{}{}
+		}
```
Comment (jsafrane) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go:390]:
```
volumesInUse = sets.New[v1.UniqueVolumeName](node.Status.VolumesInUse...)
```

And use set functions to query for existence below.

File: pkg/controller/volume/attachdetach/attach_detach_controller_test.go [hunk 2]
```diff
@@ -128,11 +127,25 @@ func Test_AttachDetachControllerStateOfWolrdPopulators_Positive(t *testing.T) {
 
 	for _, node := range nodes {
 		nodeName := types.NodeName(node.Name)
+		inUseVolumes := map[v1.UniqueVolumeName]struct{}{}
```
Comment (jsafrane) [inline pkg/controller/volume/attachdetach/attach_detach_controller_test.go:130]: Please use `sets`.

**Comment by jsafrane** [comment]
This is really a good refactoring!

**Comment by huww98** [comment]
@xing-yang @jsafrane Please help me add the milestone and lgtm label. Thanks.

**Comment by xing-yang** [comment]
@jsafrane Is this good to go?

**Pull request merged**